devices/src/pvmemcontrol.rs

```diff
@@ -0,0 +1,819 @@
+    ) -> Self {
+        // access to this address is checked
+        let buf_phys_addr = GuestAddress(buf_phys_addr.into());
+        if !guest_memory.memory().check_range(
+            buf_phys_addr,
+            std::mem::size_of::<PvmemcontrolResp>().max(std::mem::size_of::<PvmemcontrolReq>()),
+        ) {
+            warn!("guest sent invalid phys addr {:#x}", buf_phys_addr.0);
+            return PvmemcontrolDevice::new(
+                PvmemcontrolTransport::error(),
+                PvmemcontrolState::Broken,
+            );
+        }
+
+        let conn = {
+            // find an available port+byte combination, and fail if full
+            let mut next_conn = state.next_conn;
+            while state.port_buf_map.contains_key(&next_conn) {
+                next_conn = next_conn.next();
+                if next_conn == state.next_conn {
+                    warn!("connections exhausted");
+                    return PvmemcontrolDevice::new(
+                        PvmemcontrolTransport::error(),
+                        PvmemcontrolState::Broken,
+                    );
+                }
+            }
+            next_conn
+        };
+        state.next_conn = conn.next();
+        state.port_buf_map.insert(conn, buf_phys_addr);
+
+        // inform guest of the connection
+        let response = PvmemcontrolTransport::register_response(conn.command);
+
+        PvmemcontrolDevice::new(response, PvmemcontrolState::PercpuInit(state))
+    }
+
+    fn reset() -> Self {
+        PvmemcontrolDevice::new(
+            PvmemcontrolTransport::ack(),
+            PvmemcontrolState::PercpuInit(PercpuInitState::new()),
+        )
+    }
+
+    fn error() -> Self {
+        PvmemcontrolDevice::new(PvmemcontrolTransport::error(), PvmemcontrolState::Broken)
+    }
+
+    fn ready(PercpuInitState { port_buf_map, .. }: PercpuInitState) -> Self {
+        PvmemcontrolDevice::new(
+            PvmemcontrolTransport::ack(),
+            PvmemcontrolState::Ready(port_buf_map),
+        )
+    }
+
+    fn run_command(
+        &mut self,
+        guest_memory: &GuestMemoryAtomic<GuestMemoryMmap<AtomicBitmap>>,
+        command: PvmemcontrolTransportCommand,
+    ) {
+        let state = std::mem::replace(&mut self.state, PvmemcontrolState::Broken);
+
+        *self = match command {
+            PvmemcontrolTransportCommand::Reset => Self::reset(),
+            PvmemcontrolTransportCommand::Register => {
+                if let PvmemcontrolState::PercpuInit(state) = state {
+                    // SAFETY: By device protocol. If driver is wrong the device
+                    // can enter a Broken state, but the behavior is still sound.
+                    Self::register_percpu_buf(guest_memory, state, unsafe {
+                        self.transport.as_register()
+                    })
+                } else {
+                    warn!("received register without reset");
```

Comment:
If you want to defend against a buggy or malicious guest from flooding your log file the `warn!` should be changed to `debug!`.

Reply:
Makes sense. I switched the device connectivity warns to debug, but for more specific errors I kept the warns.